New design (and structure)

manuelmeister opened the pull request:
I've tried to make the design more legible and tried to make it more noob friendly. I used the old vitepress theme as a base. 

![image](https://user-images.githubusercontent.com/85400359/202273071-7d1ad039-99db-4021-aff3-548f993fa54e.png)


Preview: https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/

## Content (this will be discussed and changed after the new design is deployed)
- Made Home like a landing page for Devs 
- Add FusionPen Examples
- Restrucutred Migrations
- API Fusion Syntax Examples
- Installation Guide restructured to focus on us cases and not operating systems and use cases
  - System Requirements updated

## Improvements
- Improved visual hierarchy in guide
- Typography more focused on reading docs
- Multiple different sidebars (based on top level)
- Made Tutorials more readable and added sidebar "filter"
- Dynamic navbar
- New Fusion Syntax Highlighter that also better matches `include:` paths
- Syntax Highlighter Eel in Fusion is rendered like JS and highlights important variables like props,…
- Improve Pagespeed

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Public/Frontend/js/prism-afx.js:
```diff
@@ -0,0 +1,199 @@
+
+	spread = re(spread).source;
+
+	Prism.languages.afx = Prism.languages.extend('markup', javascript);
```
wow well done ^^

new prism AFX hightlighting

crydotsnake commented:
The preview instance is not working anymore 😕

manuelmeister commented:
Sorry my trial at @flownative ran out.

kdambekalns commented:
> Sorry my trial at @flownative ran out.

You could have asked for stage instance to be set up in the Neos project… 😜

robertlemke commented:
@manuelmeister I configured the new Beach instance so that the free trial does not expire.

manuelmeister commented:
@robertlemke @kdambekalns sorry, ich hatte es auch vergessen gehabt und habe jetzt halt noch einmal eine neue Instanz gemacht: https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/

jonnitto commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/Sidebar/Sidebar.LevelRenderer.fusion:
```diff
@@ -0,0 +1,35 @@
+        active.@if.isActive = ${node == documentNode || String.indexOf(q(node).property('target'), documentNode.identifier) > -1}
+    }
+    navigationClass = "sidebar-links"
+    navigationItems = Neos.Fusion:Collection {
+        @if.notEmpty = ${this.collection.count() > 0}
+        collection = ${q(node).children('[instanceof Neos.Neos:Document][_hidden != TRUE][_hiddenInIndex != TRUE][hideFromMainMenu != TRUE]')}
```
```suggestion
    navigationItems = Neos.Fusion:Loop {
        @if.notEmpty = ${this.items.count() > 0}
        items = ${q(node).children('[instanceof Neos.Neos:Document][_hidden != TRUE][_hiddenInIndex != TRUE][hideFromMainMenu != TRUE]')}
```

jonnitto commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Content/Tabs/Tabs.fusion:
```diff
@@ -1,18 +1,15 @@
-            <Neos.Neos:ContentCollection nodePath="" attributes.class="tabs-content col s12" />
-        </div>
+        <ul class="tabs">
+            <Neos.Fusion:Collection collection={q(node).children()} iteratorName="iterator" itemName="item">
+                <li @path="itemRenderer" class="tab col s3">
+                    <Carbon.Link:Link link={'#' + item.identifier} additionalClass={iterator.isFirst ? 'active' : ''} backendLink={true} >
+                        <Neos.Neos:Editable node={item} property="title" />
+                    </Carbon.Link:Link>
+                </li>
+            </Neos.Fusion:Collection>
```
```suggestion
            <Neos.Fusion:Loop items={q(node).children()}>
                <li class="tab col s3">
                    <a href={'#' + item.identifier} class={iterator.isFirst ? 'active' : null}>
                        <Neos.Neos:Editable node={item} property="title" />
                    </a>
                </li>
            </Neos.Fusion:Loop>
```

jonnitto reviewed (changes requested):
First review, more will follow, but hey! thanks for your work ❤️

manuelmeister commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/NavBar/NavBarItemDialog.fusion:
```diff
@@ -0,0 +1,22 @@
+            </button>
+
+            <ul class="dialog">
+                <Neos.Fusion:Collection collection={props.navigationItems} itemName="node">
+                    <li class="dialog-item">
+                        <Neos.DocsNeosIo:Component.NavBarItemDialogItem node={node} />
+                    </li>
+                </Neos.Fusion:Collection>
```
Will do 👍  

Why is it outdated?

mhsdesign replied:
because we renamed things for readability in neos 4 or 5 ...

> The difference between `RawArray`, `Array`, `RawCollection` and `Collection` was often hard to get for new developers. To overcome this, the old confusing names are deprecated, and new prototypes are introduced which are easier to understand and emphasize the declarative nature of Fusion.

https://github.com/neos/neos-development-collection/pull/2189

manuelmeister commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/Header/Header.fusion:
```diff
@@ -1,38 +1,54 @@
+    }
+
+    # rendering
+    @context {
+        startingPoint = ${this.startingPoint}
+        node = ${this.startingPoint}
+        renderCurrentPath = ${this.renderCurrentPath}
```
I don't know how to set this up without. I don't understand why it is used here.

mhsdesign replied:
see https://github.com/neos/neos-development-collection/pull/3943#issuecomment-1298259920 for alternatives ;)

manuelmeister replied:
As I copied this from [the old navigation](https://github.com/neos/Neos.DocsNeosIo/pull/95/files/2a44826668b2f87e5b82a8baa3a47b843f43b880#diff-bfddf49c0f1a60127f93a88f864c279725b9c79f7a4be777262148f5fc3d60a7R5-R11). I think, this should be resolved in a different PR alongside carbon link replacement.

manuelmeister replied:
Created #105

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/NodeTypes/Content/ApiSection.yaml:
```diff
@@ -0,0 +1,30 @@
+'Neos.DocsNeosIo:Content.ApiSection':
+  label: "${'H' + (q(node).parents('[instanceof Neos.DocsNeosIo:Content.Section]').count()+2) + ' ' + (String.cropAtWord(String.trim(String.stripTags(String.pregReplace(String.replace(q(node).property('title') || ((I18n.translate(node.nodeType.label) || node.nodeType.name) + (node.autoCreated ? ' (' + node.name + ')' : '')), '||', ''), '/<br\\W*?\\/?>|\\x{00a0}|[[^:print:]]|\\s+/u', ' '))), 100,  ...
```
oooo boi thats some fun stuff there :P

if it is necassary or worth it, i would suggest to write an own eel helper and put the logic in php - this increases readability a lot.
Eel is not perfect for this.

mhsdesign replied:
I mean i like the logic, its just neither performant nor readable in eel :P Was this also before like this? I could write this in php if you like :P

manuelmeister commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/NavBar/NavBarItemDialogItem.fusion:
```diff
@@ -0,0 +1,23 @@
+    isExternal = false
+    renderer = afx`
+        <div class="nav-dropdown-link-item">
+            <Carbon.Link:Link title={props.itemTitle} node={node} additionalClass={props.itemClass} backendLink={true}>
+                <span class="arrow"></span>
+                <span class="text">{ props.itemTitle }</span>
+                <span class="icon"><Neos.DocsNeosIo:Component.OutboundLinkIcon @if={props.isExternal}/>
+                </span>
+            </Carbon.Link:Link>
```
Okay, should I remove carbon link from the old stuff as well?

manuelmeister replied:
I did not manage to replace it with Neos.Neos:NodeLink

mhsdesign replied:
we can delay this task ;) (Open an issue with "Technical debt")

manuelmeister replied:
https://github.com/neos/Neos.DocsNeosIo/issues/105

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Document/SearchResultsPage/SearchResultsPage.fusion:
```diff
@@ -1,4 +1,15 @@
 prototype(Neos.DocsNeosIo:Document.SearchResultsPage) < prototype(Neos.DocsNeosIo:Document.AbstractPage) {
+    bodyTag.attributes.class.@process.setNoSidebar = Neos.Fusion:Case {
```
mit neos 8 kann man es auf `${Array.push(value, "no-sidebar")}` minifizieren ;)

manuelmeister replied:
Wie einfach ist das Upgrade?

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/Header/Header.fusion:
```diff
@@ -1,38 +1,54 @@
+        </div>
+
+        <div class="nav-bar">
+            <button class="sidebar-button" onclick="document.querySelector('.sidebar')?.classList.toggle('open')">
```
i sense a little alpine like js here :D

mhsdesign commented:
Wow :heart:

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Document/HomePage/HomePage.fusion:
```diff
@@ -1,13 +1,50 @@
+                    <div style="margin-top: 20px;display: grid; grid-template-columns: 1fr 1fr; gap: 20px">
+                        <Carbon.Link:Link link="/guide/getting-started" additionalClass="btn secondary">Get started
+                        </Carbon.Link:Link>
+                        <Carbon.Link:Link link="/guide/installation-development-setup" additionalClass="btn secondary">
```
hard links ...

manuelmeister commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/Layout/Layout.Default.fusion:
```diff
@@ -35,60 +37,58 @@ prototype(Neos.DocsNeosIo:Component.Layout.Default) < prototype(Neos.Neos:Conten
+            </main>
 
-            <div class="container fullwidth-layout" @if.is={props.fullwidth}>
+            <main class="container fullwidth-layout" @if.is={props.fullwidth}>
```
Should this generally be removed in afx?

mhsdesign replied:
yes - its a newer feature of afx (7.2 i think) as the `.is` doesnt add much detail.
If you feel the need to explain an expression feel free and keep the name ;)

markusguenther commented:
I am on an 8.1 update and when this is done I hope that the rebase is not so hard...

rolandschuetz commented:
I like the general look, the way the content looks and how example and tabs look.

A few minor things to improve:
 - The buttons and linklist looks a bit sloppy. The shadow animations don't really fit the Neos brand.
 - "Installation With DDEV / Local Beach" is a weird clustering, I think each on should have it's own menu item
 - "Installation Manual Setup" doesn't follow the above introduced structure, maybe "Installation On your Locale System"
 -  The tutorial pages are broken https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/tutorials/flush-varnish-cache-after-successful-surf-deployment
 - I don't understand why Tutorials is so far removed from the docs, and what the menu item "Guides" should be
 - The search is currently broken https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/search?search=fusion and all images are missing, guess this is just a setup thing and should be ignored?
 - I like the Fusion Pen examples page!
  - I think the way to actually find neos.readthedocs.io is too far, it's hidden under API References https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/api/references

manuelmeister commented:
Thanks for the review! 😄 

---

> The buttons and linklist looks a bit sloppy. The shadow animations don't really fit the Neos brand.

What do you mean with "sloppy"? Just the shadow?

---

> "Installation With DDEV / Local Beach" is a weird clustering, I think each on should have it's own menu item
>
> "Installation Manual Setup" doesn't follow the above introduced structure, maybe "Installation On your Locale System"

I just thought that from a beginner perspective, there are several paths:
- Recommended way (Docker)
- A way with helpers such as DDEV or Beach (semi recommended)
- A manual way (advanced)

But I think this should be discussed in a next step (as this PR does not include Structure changes)

---

> The tutorial pages are broken https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/tutorials/flush-varnish-cache-after-successful-surf-deployment

I think your review coincided with a deployment

---

> I don't understand why Tutorials is so far removed from the docs, and what the menu item "Guides" should be

- Tutorials are a step by step guide how to do something specific using a simple real-life example. Often I find tutorials via the google search
- Guides are more a handbook how everything is done, going from core concepts / models / priciples to advanced features. Something like the structure I did in[ my first draft](https://manuelmeister.github.io/neos-docs/guide/introduction.html).

---


> The search is currently broken https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/search?search=fusion and all images are missing, guess this is just a setup thing and should be ignored?

I don't know how to set it up on beach. On my machine it works perfectly. Maybe @kdambekalns can help here.

---

> I like the Fusion Pen examples page!

Thank you. I think code wise it needs some work.

---

> I think the way to actually find neos.readthedocs.io is too far, it's hidden under API References https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/api/references

I think for the meantime (until we integrate the code reference from github) I think we could make it, so that references is the main page when you click on api. Not a direct link to the readthedocs, but the overview page: https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/api/references

manuelmeister commented:
@rolandschuetz Thanks for the review! 😊

---

> The buttons and linklist looks a bit sloppy. The shadow animations don't really fit the Neos brand.

What do you mean with "sloppy"? Just the shadow?

---

> "Installation With DDEV / Local Beach" is a weird clustering, I think each on should have it's own menu item
>
> "Installation Manual Setup" doesn't follow the above introduced structure, maybe "Installation On your Locale System"

I just thought that from a beginner perspective, there are several paths:
- Recommended way (Docker)
- A way with helpers such as DDEV or Beach (semi recommended)
- A manual way (advanced)

But I think this should be discussed in a next step (as this PR does not include Structure changes)

---

> The tutorial pages are broken https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/tutorials/flush-varnish-cache-after-successful-surf-deployment

I think your review coincided with a deployment

---

> I don't understand why Tutorials is so far removed from the docs, and what the menu item "Guides" should be

- Tutorials are a step by step guide how to do something specific using a simple real-life example. Often I find tutorials via the google search
- Guides are more a handbook how everything is done, going from core concepts / models / priciples to advanced features. Something like the structure I did in[ my first draft](https://manuelmeister.github.io/neos-docs/guide/introduction.html).

---


> The search is currently broken https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/search?search=fusion and all images are missing, guess this is just a setup thing and should be ignored?

I don't know how to set it up on beach. On my machine it works perfectly. Maybe @kdambekalns can help here.

---

> I like the Fusion Pen examples page!

Thank you. I think code wise it needs some work.

---

> I think the way to actually find neos.readthedocs.io is too far, it's hidden under API References https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/api/references

I think for the meantime (until we integrate the code reference from github) I think we could make it, so that references is the main page when you click on api. Not a direct link to the readthedocs, but the overview page: https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/api/references

manuelmeister commented:
I did not want to remove the request for review

rolandschuetz commented:
It's a little hard to explain, or a lot of work to define "sloppy", some elements just don't have the CI-feel to it. E.g. probably LinkList.Items should not be bold, since they are normal content. Hover effects should have transitions, some spacings look a bit off... it's confusing that some headlines have a direct link and some don't (e.g. [Read API on homepahe](https://instance-7602d6b0-3aac-4c9a-bd63-6f382cbf56aa.flownative.cloud/#read-api) has a link, but Ask the community not), link hover effects should make the element light blue like on our website, and not add a underline for elements like lists or buttons, ...

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/LocalNavigation/LocalNavigation.fusion:
```diff
@@ -9,7 +9,7 @@ prototype(Neos.DocsNeosIo:Component.LocalNavigation) < prototype(Neos.Fusion:Com
         <div class="local-navigation sm-hidden">
             <aside class="toc">
-                <Neos.DocsNeosIo:Component.LocalNavigation.LevelRenderer collection={props.collection}/>
+                <Neos.DocsNeosIo:Component.LocalNavigation.LevelRenderer items={props.collection}/>
```
das sieht nicht ganz koscher aus ... die api sagt es heißt nun `items` ...

ich würde im IDE nochmal nach dem string `collection` in allen fusion datein suchen ;)

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/TagList/TagList.fusion:
```diff
@@ -1,12 +1,12 @@
-    @context.collection = ${q(this.collection).filter(this.filter)}
-    tags = Neos.Fusion:Collection {
-        collection = ${collection}
+    @context.collection = ${q(this.items).filter(this.filter)}
```
huch müsste das auch nicht items sein? `@context.collection`

manuelmeister commented:
@rolandschuetz I've tried to make the whole docs design as accessible as possible. The neos color sadly is really bright and does not combine well with white and is not very readable. If on hover it is even brighter, it makes the contrast even worse.

> **Contrast**
> WCAG: https://webaim.org/resources/contrastchecker/?fcolor=00ADEE&bcolor=FFFFFF
> APCA: https://www.myndex.com/APCA/?BG=ffffff&TXT=00adee&DEV=G4g&BUF=A22

I thought, in general it would make sense to diverge from the branding CI (I very like for print products) and make it more accessible. To keep the brand color #00ADEE for links instead of a darker shade, I've made the links slightly bolder (medium) and added an underline on hover, as this is universally understandable as a link. I think that dev docs need to be very readable and have an easy to understand visual hierarchy.

I agree that the shadow is not very "on brand". I'll rework this part. Also for the hover effect I think I'll find an improved but consistent way.

This has not changed with my PR. The linklist title was not hashlinked before. I've just replaced the custom html widget with sections. @markusguenther said that he wants to remove the widget anyway: https://github.com/neos/Neos.DocsNeosIo/pull/100#issuecomment-1331980822 

If you want, we can discuss the design bilaterally in a phone call. I find written design feedback difficult because then a lot of nuances get lost and it's usually not so clear to me how to weight the different statements.

manuelmeister commented:
With the current design trends, creating accessible design is hard.

mhsdesign commented on DistributionPackages/Neos.DocsNeosIo/Resources/Private/Fusion/Component/Sidebar/Sidebar.scss:
```diff
@@ -0,0 +1,255 @@
+    padding-left: 2rem
+}
+
+.sidebar>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-link-item,.sidebar>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-links>.sidebar-link>.sidebar-link-item {
```
öhm :D

manuelmeister replied:
Wenn du willst, kann ich diesen Code von Vitepress mit scss vereinfachen.

mhsdesign replied:
ich kann mir aber nicht vorstellen, dass das lesbarer werden kann ^^

hmm gut das ist ja kopiert, also waren einfach die vitepress leute ein bisschen wahnsinnig ^^ 
Wenn man sich den selektor durchliest muss man einfach nur schmunzeln - aufräumen kann (oder nicht 😅) man ja sonst auch später :D

mhsdesign replied:
thanks ❤️ looks better ;)

manuelmeister commented on DistributionPackages/Neos.DocsNeosIo/NodeTypes/Content/ApiSection.yaml:
```diff
@@ -0,0 +1,30 @@
+    'Neos.DocsNeosIo:Constraint.Content.LocalNavigation.Item': true
+  ui:
+    label: 'Api Section'
+    icon: icon-level-down
```
Added todo to change all `icon-` prefixes: #105

manuelmeister commented:
@jonnitto thank you for the review! I've either fixed those issues or added them to the cleanup task (#105). Could you take another look?

manuelmeister commented:
## Meeting with @rolandschuetz

### Next steps
1. Implement change requests from @rolandschuetz 
2. Wait for 8.2 update #100 and review by @Sebobo 
3. Test on [stage](https://instance-6ee89105-ec8e-4024-97bb-5c2b9d49a15c.flownative.cloud/)
4. Deployment by @rolandschuetz

rolandschuetz reviewed (changes requested):
A few changes

manuelmeister commented on DistributionPackages/Neos.DocsNeosIo/NodeTypes/Content/HTML.yaml:
```diff
@@ -0,0 +1,23 @@
+'Neos.DocsNeosIo:Content.HTML':
```
Renamed as discussed in favor of a CodePreview node

manuelmeister commented:
@rolandschuetz I've implemented the requested changes.

markusguenther commented:
Seems that I am not allowed to update the branch ;)

manuelmeister commented:
Fixed it

manuelmeister commented:
@markusguenther fixed it :)
